### DISCUSSION TITLE
Add options to top level command

### DISCUSSION THREAD

**Discussion by dariober** [oclif/core#954] [Ideas]
Do you want to request a *feature* or report a *bug*?
-----------------------------------------------------

Most likely this is a question.

----

I'd like to add options to the top level command, similar to *e.g.* `git` where you can do:

```
git --paginate log
git --paginate show
```

and the option `--paginate` is passed to subcommands log and show.

From the [hello-world](https://github.com/oclif/hello-world/tree/main) example I can see how to add options to subcommands by editing the index.ts files (e.g. `src/commands/hello/index.ts`). However, I'm not able to do the same for the top level command. I tried to tweak `src/index.ts` and the package settings, but I had no luck so far.

**Comment by mdonnalley** [comment]
This isn't currently supported since oclif expects all flags and args to be placed after the command id. Although I think we could likely support it if there's enough demand for it (or someone would like to make a PR).

I'm imagining some sort of configuration like this

```json
"oclif": {
  "preCommandFlags": ["paginate"]
}
```

Then this [function](https://github.com/oclif/core/blob/3.19.1/src/help/util.ts#L98) (and the handful of functions it invokes) would need to know that the command id will be found after all the `preCommandFlags`

As for defining and parsing the `preCommandFlags` - I think that would be the responsibility of the developers would want to use them. Meaning, that oclif wouldn't do anything magical and you'd still need to define the flag on every command that wants to use it.

So putting it all together, you'd have this configuration in your package.json

```json
"oclif": {
  "preCommandFlags": ["paginate"]
}
```

Then you'd have a [base command](https://oclif.io/docs/base_class) that defines the `paginate` flag. All commands would then need to extend this base class in order to inherit the `paginate` flag

That's just off the top of my head though - I'm open to alternative implementations
